src/amd64/amd64.c.inc

```diff
@@ -0,0 +1,201 @@
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Parts of this file from biench.cr.yp.to/supercop.html (2017-02-25):
+//     Daniel J. Bernstein
+//     Niels Duif
+//     Tanja Lange
+//     lead: Peter Schwabe
+//     Bo-Yin Yang
+
+#include <string.h>
+#include "ge25519.h"
+
+extern void choose_tp(ge25519_pniels *t, unsigned long long pos, signed long long b, const ge25519_pniels *base_multiples);
+
+/* return 0 on success, -1 otherwise. Taken from
+ed25519/amd64-51-30k/ge25519_unpackneg.c - the negation is removed. */
+static int unpack_vartime(ge25519_p3 *r, const unsigned char p[32])
+{
+  fe25519 t, chk, num, den, den2, den4, den6;
+  unsigned char par = p[31] >> 7;
+
+  fe25519_setint(&r->z,1);
+  fe25519_unpack(&r->y, p); 
+  fe25519_square(&num, &r->y); /* x = y^2 */
+  fe25519_mul(&den, &num, &ecd); /* den = dy^2 */
+  fe25519_sub(&num, &num, &r->z); /* x = y^2-1 */
+  fe25519_add(&den, &r->z, &den); /* den = dy^2+1 */
+
+  /* Computation of sqrt(num/den)
+     1.: computation of num^((p-5)/8)*den^((7p-35)/8) = (num*den^7)^((p-5)/8)
+  */
+  fe25519_square(&den2, &den);
+  fe25519_square(&den4, &den2);
+  fe25519_mul(&den6, &den4, &den2);
+  fe25519_mul(&t, &den6, &num);
+  fe25519_mul(&t, &t, &den);
+
+  fe25519_pow2523(&t, &t);
+  /* 2. computation of r->x = t * num * den^3
+  */
+  fe25519_mul(&t, &t, &num);
+  fe25519_mul(&t, &t, &den);
+  fe25519_mul(&t, &t, &den);
+  fe25519_mul(&r->x, &t, &den);
+
+  /* 3. Check whether sqrt computation gave correct result, multiply by sqrt(-1) if not:
+  */
+  fe25519_square(&chk, &r->x);
+  fe25519_mul(&chk, &chk, &den);
+  if (!fe25519_iseq_vartime(&chk, &num))
+    fe25519_mul(&r->x, &r->x, &sqrtm1);
+
+  /* 4. Now we have one of the two square roots, except if input was not a square
+  */
+  fe25519_square(&chk, &r->x);
+  fe25519_mul(&chk, &chk, &den);
+  if (!fe25519_iseq_vartime(&chk, &num))
+    return -1;
+
+  /* 5. Choose the desired square root according to parity:
+  */
+  if(fe25519_getparity(&r->x) == (1-par)) // only change from original function was `!=` -> `==`
+    fe25519_neg(&r->x, &r->x);
+
+  fe25519_mul(&r->t, &r->x, &r->y);
+  return 0;
+}
+
+static void p3_to_pniels(ge25519_pniels* out, ge25519_p3 const* src)
+{
+  fe25519_sub(&out->ysubx, &src->y, &src->x);
+  fe25519_add(&out->xaddy, &src->x, &src->y);
+  fe25519_mul(&out->t2d, &src->t, &ec2d);
+  out->z = src->z;
+}
+
+static void negate(ge25519* out)
+{
+  fe25519_neg(&out->x, &out->x);
+  fe25519_neg(&out->t, &out->t);
+}
+
+// similar to loops in existing implementation, but uses dynamic table instead of fixed `G`.
+static void monero_scalarmult(ge25519_p3* r, ge25519_pniels const* base, char const* sec)
```

Comment:
Maybe an unfortunate name since it does a mul8 at the end. static so does not matter that much I guess.

Reply:
I named it `monero_scalarmult` because its not-standard. But you are correct, this clashes with the naming scheme elsewhere. `scalarmult_cryptonote` ?

Reply:
I was kinda thinking just add "8" or something along those lines.

Reply:
Seems most similar to https://github.com/monero-project/monero/blob/master/src/crypto/crypto.cpp#L184

Reply:
Renamed, see updated version being pushed shortly.